### PR TITLE
Fix `toe` command

### LINKED ISSUE #1642: toe does not load existing type in editor and it does not change it afterward
| Questions                                            | Answers
|------------------------------------------------------|--------------------
| OS/arch/bits (mandatory)                             | Fedora 33
| File format of the file you reverse (mandatory)      | PE, ELF etc.
| Architecture/bits of the file (mandatory)            | x86_64
| `rizin -v` full output, **not truncated** (mandatory)  | rizin 0.3.0-git

### Expected behavior
Current c output of the given type is automatically added in the buffer and the new value is correctly parsed.

### Actual behavior
Vim is executed, but `tm` structure is not loaded in the vim buffer. Also, when you save, there's a parser error even if the data is right.

### Steps to reproduce the behavior
```
$ rizin 
> toe tm
```

### Additional Logs, screenshots, source code,  configuration dump, ...

Few things I noticed:
1) use rz_cmd_escape_arg when executing `tc` command in `rz_types_open_editor`
2) use `rz_core_editor(core, NULL, str)` instead of `*.h`. It seems that somehow that fixes the first problem, that is that data is not loaded in the vim buffer.
3) even though i tried the step 2), something failed while parsing the type back from vim. I'm not sure why.

### PR DESCRIPTION
**Pull request by XVilka**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

Changed the `toe` implementation to accept only one type name per call, fixed bugs.

**Test plan**

Try `toe tm` for example

**Closing issues**

Closes https://github.com/rizinorg/rizin/issues/1642

### CODE REVIEW

File: librz/core/ctypes.c [hunk 1]
```diff
@@ -1004,44 +1066,68 @@ RZ_IPI bool rz_types_open_file(RzCore *core, const char *path) {
...
+	if (t->kind == RZ_BASE_TYPE_KIND_ATOMIC) {
+		RZ_LOG_ERROR("Cannot edit atomic type \"%s\"\n", typename);
+		return false;
+	}
+	char *str = rz_core_base_type_as_c(core, t, true);
+	if (!str) {
+		RZ_LOG_ERROR("Cannot genereate C representation of type \"%s\"\n", typename);
+		return false;
+	}
+	char *tmp = rz_core_editor(core, NULL, str);
```
Comment (thestr4ng3r) [inline librz/core/ctypes.c:1104]: Please add the code below as an API, so we can use it in Cutter.

**Review by wargio — APPROVED** [review]

**Review by ret2libc — CHANGES_REQUESTED** [review]
Doesn't this suffer from the problem we discussed the other day? That is that if the new type is invalid, you will end up deleting the existing type?

Few solutions:
1) save the original C string representation and re-parse it in case the new one fails
2) remove the original basetype from the hashtable, but don't free it. If the new one fails you re-insert the basetype in the hashtable

**Comment by XVilka** [comment]
@ret2libc it already does that: https://github.com/rizinorg/rizin/pull/1659/files#diff-9ad7c99ed0a1b70e3fd759554220a7ff325910afc935d78fb8c342387de91e81R1119

File: librz/core/ctypes.c [hunk 2]
```diff
@@ -1004,44 +1066,99 @@ RZ_IPI bool rz_types_open_file(RzCore *core, const char *path) {
...
+	char *error_msg = NULL;
+	int result = rz_type_parse_string_stateless(typedb->parser, new, &error_msg);
+	if (result) {
 		if (error_msg) {
-			eprintf("%s\n", error_msg);
+			RZ_LOG_ERROR("%s\n", error_msg);
+		}
+		free(error_msg);
+		// There is an error during the parsing thus we restore the old type
+		int result = rz_type_parse_string_stateless(typedb->parser, old, &error_msg);
```
Comment (thestr4ng3r) [inline librz/core/ctypes.c]: What about not using `rz_type_db_delete_base_type()` to delete the base type entirely, but just removing it from the hashtable without freeing to re-insert it later?
Using the `old` string means that it must be guaranteed that printing a type to C and then parsing it again will always result in exactly the original type, which I would not trust if not strictly necessary.

**Comment by XVilka** [comment]
@thestr4ng3r @ret2libc added new API function:
```c
/**
 * \brief Edits the existing base type given the new C code
 *
 * Searches the base type in the types database given the \p name.
 * If it exists - parses the \p typestr as the new C type. If there is
 * any error during the parsing it restores the original type in the
 * database.
 *
 * \param typedb Type Database instance
 * \param name Name of the base type
 * \param typestr C string of the new definition of the type
 */
RZ_API bool rz_type_db_edit_base_type(RzTypeDB *typedb, RZ_NONNULL const char *name, RZ_NONNULL const char *typestr)
```

**Review by thestr4ng3r — APPROVED** [review]
Perfect!

**Pull request merged**